### DISCUSSION TITLE
BitcoinJS integration

### DISCUSSION THREAD

**Discussion by junderw** [rust-bitcoin/rust-bitcoin#1738] [General]
Since moving our tiny-secp256k1 library to use WASM through Rust, I have been playing around with the idea of creating a thin wrapper over rust-bitcoin and exposing it as a TypeScript API using wasm-bindgen.

Currently there are so many frameworks and bundlers in the JS ecosystem that it makes WASM support difficult. It feels like a game of whack-a-mole trying to get all the bundlers to work with it, and then you always get the occasional user that is stuck on an old version of an old bundler... or a user that wants to support IE11 (in 2023) or something crazy like that.

Also, looking at the state of rust-bitcoin, it seems like things are still in flux until a v1 release.

This thread is just to guage opinions on how Rust (specifically rust-bitcoin) will play into the JS/TS ecosystem in the future, and how BitcoinJS can work with rust-bitcoin to help amplify each others' work.

This is probably not a 2023 (or even 2024) goal, but at some point I'd like to start experimenting.

**Comment by apoelstra** [comment]
We have, in the past, run WASM in our CI. It looks like right now we're doing this for `hashes` only. We'd be happy to extend that support. But of course, when it inevitably breaks, and upon looking into it we discover that all the tooling and crap has changed 20 times, we'll have to remove it again.

We have put a lot of effort into no_std compatibility, and WASM compatibility is one of our main motivations for this. So in short,
* We are happy to "passively" support WASM directly, by accepting contributions to improve testing, compatibility, CI, etc
* We are also happy to "actively" work on improving no_std support in general

There are a couple big issues right now:
* `no_std` support is kinda a mess, because we use the `io::Write` trait heavily and for some reason this is not in stdlib. In rust-bitcoin we also need an allocator, though this isn't true for rust-secp or hashes. We have a weird mess of feature gates and optional dependencies that we'd like to clean up but we need more ecosystem support.
* rust-secp, as you know, binds to a C library, and as near as I can tell direct WASM support for that is not on the roadmap for the Rust devs. We'd be happy to explore alternate routes, like swapping out `cc` for something that would directly compile C to WASM, and then binding to that ... but it's a lot of work and none of us are expert enough in that stuff.

**Reply by TheBlueMatt** [reply]
> no_std support is kinda a mess, because we use the io::Write trait heavily and for some reason this is not in stdlib. In rust-bitcoin we also need an allocator, though this isn't true for rust-secp or hashes. We have a weird mess of feature gates and optional dependencies that we'd like to clean up but we need more ecosystem support.

I don't think any of these are a big deal for no-std - the allocator is of course no problem, but same with no-std support generally.

> rust-secp, as you know, binds to a C library, and as near as I can tell direct WASM support for that is not on the roadmap for the Rust devs. We'd be happy to explore alternate routes, like swapping out cc for something that would directly compile C to WASM, and then binding to that ... but it's a lot of work and none of us are expert enough in that stuff.

FWIW, this means no `wasm_bindgen`, it doesnt mean any of the *other* rust wasm targets are impacted.

On the LDK end we do package LDK as a WASM + wrapper library, but not using wasm_bindgen, rather using our own home-grown bindings generation, which links using the `wasm32-wasi` target which works great.

**Comment by Kixunil** [comment]
Heh, I was thinking about suggesting slapping `#[cfg_attr(feature = "wasm-bindgen", wasm-bindgen)]` on some of our types, I think that alone would help a lot. We'll need to add constructors too but shouldn't be too bad.

I'm actually working on a wasm-using Bitcoin Rust project so this might be useful to me. So far I managed to not need this though.
One particularly annoying thing is there's no way to transfer raw bytes between JS and Rust so I had to base64-encode everything :see_no_evil:

**Reply by TheBlueMatt** [reply]
> Heh, I was thinking about suggesting slapping #[cfg_attr(feature = "wasm-bindgen", wasm-bindgen)] on some of our types

I'm still a big uneasy about saying we support wasm_bindgen, at least for anything that touches libsecp. Last I heard we *think* its safe in the sense that the issues with ABI compatibility should result in different arguments to a function which the linker should at least warn about (newer linkers fail, older ones do generate warnings, but I believe wasm_bindgen wont show you those warnings :facepalm: ), but I don't get the sense upstream cares too much about this so isn't spending much time thinking about it.

>  One particularly annoying thing is there's no way to transfer raw bytes between JS and Rust so I had to base64-encode everything see_no_evil

Lol wasm_bindgen is such a mess...what a strange restriction.

**Reply by Kixunil** [reply]
Is there some test suite I could run to verify everything is OK? This is scary as hell.

**Reply by junderw** [reply]
> no way to transfer raw bytes between JS and Rust so I had to base64-encode everything

You can pass `Uint8Array` in as an arg and take it as `&[u8]` and in return position you can return a `Uint8Array` by returning a `Box<[u8]>` unfortunately, `Vec<u8>` is not supported though. ([Issue for return value Vec support](https://github.com/rustwasm/wasm-bindgen/issues/111))

**Reply by TheBlueMatt** [reply]
> Is there some test suite I could run to verify everything is OK? This is scary as hell.

I *think* as long as you use a new (ish) linker you're okay...

**Reply by Kixunil** [reply]
Can you recommend any resources documenting how to check my linker is new enough?

**Reply by Kixunil** [reply]
@junderw oh, didn't know about `Box<[u8]>`, thanks!

**Reply by TheBlueMatt** [reply]
https://github.com/rust-lang/cc-rs/issues/715
